universum/configuration_support.py

```diff
@@ -1,18 +1,265 @@
 ]
 
-skip_attributes = {"children", "critical", "skip_numbering_level"}
 
+class ProjectConfiguration:
+    """
+    ProjectConfiguration is a collection of configurable project-specific data needed for Universum operations.
+    Legacy constructor supports dictionary data, but for static type checking please use new explicit constructor.
+    """
+
+    # pylint: disable-msg=too-many-locals
+    def __init__(self,
+                 name: str = '',
+                 command: Optional[List[str]] = None,
+                 environment: Optional[Dict[str, str]] = None,
+                 artifacts: str = '',
+                 report_artifacts: str = '',
+                 artifact_prebuild_clean: bool = False,
+                 directory: str = '',
+                 critical: bool = False,
+                 background: bool = False,
+                 finish_background: bool = False,
+                 code_report: bool = False,
+                 pass_tag: str = '',
+                 fail_tag: str = '',
+                 if_env_set: str = '',
+                 **kwargs) -> None:
+        """
+        All params supplied via named attributed shall be  type-checked for safety. Clients may provide custom
+        parameters in kwargs - these parameters will be stored internally in '_extras' dict and can be retrieved by
+        indexing.
+        >>> ProjectConfiguration(foo='bar')
+        {'foo': 'bar'}
+        >>> cfg = ProjectConfiguration(name='foo', command=['1', '2', '3'], _extras={'_extras': 'test', 'bool': True})
+        >>> cfg
+        {'name': 'foo', 'command': ['1', '2', '3'], '_extras': {'_extras': 'test', 'bool': True}}
+        >>> cfg['name']
+        'foo'
+        >>> cfg['_extras']
+        {'_extras': 'test', 'bool': True}
+        """  # TODO: move param description from configuration.rst
+        self.name: str = name
+        self.directory: str = directory
+        self.code_report: bool = code_report
+        self.command: List[str] = command if command else []
+        self.environment: Dict[str, str] = environment if environment else {}
+        self.artifacts: str = artifacts
+        self.report_artifacts: str = report_artifacts
+        self.artifact_prebuild_clean: bool = artifact_prebuild_clean
+        self.critical: bool = critical
+        self.background: bool = background
+        self.finish_background: bool = finish_background
+        self.pass_tag: str = pass_tag
+        self.fail_tag: str = fail_tag
+        self.if_env_set: str = if_env_set
+        self.children: Optional['Variations'] = None
+        self._extras: Dict[str, str] = {}
+        for key, value in kwargs.items():
+            self._extras[key] = value
+
+    def __repr__(self) -> str:
+        """
+        This function simulates dict-like representation for string output. This is useful when printing contents of
+        :class:`.Variations` objects, as internally they wrap a list of :class:`.ProjectConfiguration` objects
```

Comment:
As I understand it, :class:`Variations` without dot in the beginning should work as well, since this is the very same module. And it kinda looks better in my opinion.